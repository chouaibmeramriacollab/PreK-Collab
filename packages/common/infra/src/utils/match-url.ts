const reWeburl =
  /\b((?:(?:https?|mailto|ftp):(?:\/{2}|[a-z0-9%])|www\d{0,3}[.]|(?:[a-z0-9-]+[.])+[a-z]{2,4}\/|localhost)(?:[^\s()<>]+|\(([^\s()<>]+|(\([^\s()<>]+\)))*\))+(?:\(([^\s()<>]+|(\([^\s()<>]+\)))*\)|[^\s`!()[\]{};:'".,<>?«»。“，【】「」《》“”‘’]))/gi;

export const matchUrl = (text: string) => {
  return text.matchAll(reWeburl);
};
