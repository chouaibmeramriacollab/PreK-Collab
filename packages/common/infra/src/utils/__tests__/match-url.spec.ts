import { expect, test } from 'vitest';

import { matchUrl } from '../match-url';

test('matchUrl', () => {
  expect([
    ...matchUrl(`
    text http://example.com text
    text https://example.com text http://example.com
    example.com localhost localhost:8080
    http://localhost:8080 example.com/foo/bar
    http://foo.com/blah_(wikipedia)#cite-1
	http://foo.com/blah_(wikipedia)_blah#cite-1
	http://foo.com/unicode_(✪)_in_parens
	http://foo.com/(something)?after=parens
    (Something like http://foo.com/blah_blah_(wikipedia))
	http://foo.com/blah_blah.
	http://foo.com/blah_blah/.
    6:00p
    filename.txt
    www.example.com
    WWW.EXAMPLE.COM
    “is.gd/foo/”
    http://example.com/something?with,commas,in,url, but not at end
    http://example.com/something?with,。,in,url。 but not at end
    http://example.com/something?with,【】,in,url【】 but not at end
    http://example.com/something?with,"",in,url"" but not at end
    <http://example.com/something?within>
    「http://example.com/something?with「」inurl」
	What about <mailto:gruber@daringfireball.net?subject=TEST> (including brokets).
    `),
  ]).toMatchSnapshot();
});
