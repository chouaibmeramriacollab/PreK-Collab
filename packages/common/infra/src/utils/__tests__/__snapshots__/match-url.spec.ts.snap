// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matchUrl 1`] = `
[
  [
    "http://example.com",
    "http://example.com",
    undefined,
    undefined,
    undefined,
    undefined,
  ],
  [
    "https://example.com",
    "https://example.com",
    undefined,
    undefined,
    undefined,
    undefined,
  ],
  [
    "http://example.com",
    "http://example.com",
    undefined,
    undefined,
    undefined,
    undefined,
  ],
  [
    "localhost:8080",
    "localhost:8080",
    undefined,
    undefined,
    undefined,
    undefined,
  ],
  [
    "http://localhost:8080",
    "http://localhost:8080",
    undefined,
    undefined,
    undefined,
    undefined,
  ],
  [
    "example.com/foo/bar",
    "example.com/foo/bar",
    undefined,
    undefined,
    undefined,
    undefined,
  ],
  [
    "http://foo.com/blah_(wikipedia)#cite-1",
    "http://foo.com/blah_(wikipedia)#cite-1",
    undefined,
    undefined,
    undefined,
    undefined,
  ],
  [
    "http://foo.com/blah_(wikipedia)_blah#cite-1",
    "http://foo.com/blah_(wikipedia)_blah#cite-1",
    undefined,
    undefined,
    undefined,
    undefined,
  ],
  [
    "http://foo.com/unicode_(✪)_in_parens",
    "http://foo.com/unicode_(✪)_in_parens",
    undefined,
    undefined,
    undefined,
    undefined,
  ],
  [
    "http://foo.com/(something)?after=parens",
    "http://foo.com/(something)?after=parens",
    undefined,
    undefined,
    undefined,
    undefined,
  ],
  [
    "http://foo.com/blah_blah_(wikipedia)",
    "http://foo.com/blah_blah_(wikipedia)",
    undefined,
    undefined,
    "wikipedia",
    undefined,
  ],
  [
    "http://foo.com/blah_blah",
    "http://foo.com/blah_blah",
    undefined,
    undefined,
    undefined,
    undefined,
  ],
  [
    "http://foo.com/blah_blah/",
    "http://foo.com/blah_blah/",
    undefined,
    undefined,
    undefined,
    undefined,
  ],
  [
    "www.example.com",
    "www.example.com",
    undefined,
    undefined,
    undefined,
    undefined,
  ],
  [
    "WWW.EXAMPLE.COM",
    "WWW.EXAMPLE.COM",
    undefined,
    undefined,
    undefined,
    undefined,
  ],
  [
    "is.gd/foo/",
    "is.gd/foo/",
    undefined,
    undefined,
    undefined,
    undefined,
  ],
  [
    "http://example.com/something?with,commas,in,url",
    "http://example.com/something?with,commas,in,url",
    undefined,
    undefined,
    undefined,
    undefined,
  ],
  [
    "http://example.com/something?with,。,in,url",
    "http://example.com/something?with,。,in,url",
    undefined,
    undefined,
    undefined,
    undefined,
  ],
  [
    "http://example.com/something?with,【】,in,url",
    "http://example.com/something?with,【】,in,url",
    undefined,
    undefined,
    undefined,
    undefined,
  ],
  [
    "http://example.com/something?with,"",in,url",
    "http://example.com/something?with,"",in,url",
    undefined,
    undefined,
    undefined,
    undefined,
  ],
  [
    "http://example.com/something?within",
    "http://example.com/something?within",
    undefined,
    undefined,
    undefined,
    undefined,
  ],
  [
    "http://example.com/something?with「」inurl",
    "http://example.com/something?with「」inurl",
    undefined,
    undefined,
    undefined,
    undefined,
  ],
  [
    "mailto:gruber@daringfireball.net?subject=TEST",
    "mailto:gruber@daringfireball.net?subject=TEST",
    undefined,
    undefined,
    undefined,
    undefined,
  ],
]
`;
